py-polars/polars/internals/series.py

```diff
@@ -2118,47 +2120,80 @@ def __array__(self, dtype: Any = None) -> np.ndarray:
             return self.to_numpy().__array__()
 
     def __array_ufunc__(
-        self, ufunc: Callable[..., Any], method: str, *inputs: Any, **kwargs: Any
+        self,
+        ufunc: np.ufunc,
+        method: str,
+        *inputs: Any,
+        **kwargs: Any,
     ) -> "Series":
         """
         Numpy universal functions.
         """
         if self._s.n_chunks() > 1:
             self._s.rechunk(in_place=True)
 
+        s = self._s
+
         if method == "__call__":
+            if not ufunc.nout == 1:
+                raise NotImplementedError(
+                    "Only ufuncs that return one 1D array, are supported."
+                )
+
             args: List[Union[Number, np.ndarray]] = []
+
             for arg in inputs:
+                # Get all possible output dtypes for ufunc.
+                # Input dtypes and output dtypes seem to always match for ufunc.types,
+                # so pick all the different output dtypes.
+                dtypes_ufunc = [
+                    input_output_type[-1]
+                    for input_output_type in ufunc.types
+                    if supported_numpy_char_code(input_output_type[-1])
+                ]
+
                 if isinstance(arg, Number):
                     args.append(arg)
+                elif isinstance(arg, np.ndarray):
+                    args.append(arg)
                 elif isinstance(arg, Series):
                     args.append(arg.view(ignore_nulls=True))
                 else:
-                    return NotImplemented
+                    raise ValueError(f"Unsupported type {type(arg)} for {arg}.")
+
+            # Get minimum dtype needed to be able to cast all input arguments to the
+            # same dtype.
+            dtype_char_minimum = np.result_type(*args).char  # type: ignore[arg-type]
```

Comment:
Mypy complains about this ignore being unnecessary on 3.7, but it is needed for 3.10. The easiest way to work around this is to drop the ignore, and use `typing.cast` instead:

```python
 dtype_char_minimum = typing.cast(str, np.result_type(*args))
```
Using `cast` tells mypy that it should assume the variable is of the dtype provided, and use that for the remainder of the code.

(assuming it is a string, havent run the code myself)

Reply:
it was a problem with `*args` itself.